File: python/pia/resources/market.py
"""Official PIA SDK - Market Intelligence Resource."""

from __future__ import annotations

from typing import Any, Dict, Optional, Union
from urllib.parse import quote

from ..errors import ValidationError
from ..transport import AsyncTransport, SyncTransport
from ..types import CandleResponse, MarketPricesResponse, OrderBook


def _validate_symbol(symbol: str) -> str:
    if not isinstance(symbol, str) or not symbol.strip():
        raise ValidationError("Symbol must be a non-empty string.", param_name="symbol")
    return symbol.strip().upper()


class MarketResource:
    """Synchronous Market Data API resource."""

    def __init__(self, transport: SyncTransport) -> None:
        self._transport = transport

    def get_prices(self) -> MarketPricesResponse:
        """Retrieves a live price snapshot for all tracked global financial instruments."""
        data = self._transport.request("/api/v1/market/prices", method="GET")
        return MarketPricesResponse.from_dict(data)

    def get_candles(
        self,
        symbol: str,
        *,
        timeframe: str = "1h",
        limit: Optional[int] = None,
        since: Optional[int] = None,
        until: Optional[int] = None,
    ) -> CandleResponse:
        """Fetches historical candlestick bars for a specific instrument and timeframe."""
        clean_symbol = _validate_symbol(symbol)
        params: Dict[str, Any] = {"resolution": timeframe}
        if limit is not None:
            params["limit"] = limit
        if since is not None:
            params["since"] = since
        if until is not None:
            params["before"] = until

        endpoint = f"/api/v1/market/history/{quote(clean_symbol)}"
        data = self._transport.request(endpoint, method="GET", params=params)
        return CandleResponse.from_dict(data, symbol=clean_symbol, timeframe=timeframe)

    def get_order_book(self, symbol: str) -> OrderBook:
        """Retrieves Level 2 Depth of Market (DOM) order book for an instrument."""
        clean_symbol = _validate_symbol(symbol)
        endpoint = f"/api/v1/market/orderbook/{quote(clean_symbol)}"
        data = self._transport.request(endpoint, method="GET")
        return OrderBook.from_dict(data)


class AsyncMarketResource:
    """Asynchronous Market Data API resource."""

    def __init__(self, transport: AsyncTransport) -> None:
        self._transport = transport

    async def get_prices(self) -> MarketPricesResponse:
        """Retrieves a live price snapshot for all tracked global financial instruments."""
        data = await self._transport.request("/api/v1/market/prices", method="GET")
        return MarketPricesResponse.from_dict(data)

    async def get_candles(
        self,
        symbol: str,
        *,
        timeframe: str = "1h",
        limit: Optional[int] = None,
        since: Optional[int] = None,
        until: Optional[int] = None,
    ) -> CandleResponse:
        """Fetches historical candlestick bars for a specific instrument and timeframe."""
        clean_symbol = _validate_symbol(symbol)
        params: Dict[str, Any] = {"resolution": timeframe}
        if limit is not None:
            params["limit"] = limit
        if since is not None:
            params["since"] = since
        if until is not None:
            params["before"] = until

        endpoint = f"/api/v1/market/history/{quote(clean_symbol)}"
        data = await self._transport.request(endpoint, method="GET", params=params)
        return CandleResponse.from_dict(data, symbol=clean_symbol, timeframe=timeframe)

    async def get_order_book(self, symbol: str) -> OrderBook:
        """Retrieves Level 2 Depth of Market (DOM) order book for an instrument."""
        clean_symbol = _validate_symbol(symbol)
        endpoint = f"/api/v1/market/orderbook/{quote(clean_symbol)}"
        data = await self._transport.request(endpoint, method="GET")
        return OrderBook.from_dict(data)

File: typescript/src/resources/market.ts
/**
 * Official PIA SDK - Market Intelligence API Resource
 */

import { normalizeSymbol } from "../symbols";
import type { HttpTransport } from "../http/transport";
import type {
  CandleResponse,
  GetCandlesOptions,
  GetSymbolsOptions,
  MarketPricesResponse,
  MarketPrice,
  OrderBook,
  RequestOptions,
  SymbolsResponse,
} from "../types";

export class MarketResource {
  constructor(private readonly transport: HttpTransport) {}

  /**
   * Retrieves the catalog of all supported and tradable market instruments with precision & exchange metadata.
   */
  public async getSymbols(options?: GetSymbolsOptions): Promise<SymbolsResponse> {
    const params = new URLSearchParams();
    if (options?.asset_type) params.set("asset_type", options.asset_type);
    if (options?.search) params.set("search", options.search);
    if (options?.exchange) params.set("exchange", options.exchange);
    if (options?.limit !== undefined) params.set("limit", String(options.limit));
    if (options?.offset !== undefined) params.set("offset", String(options.offset));

    const query = params.toString() ? `?${params.toString()}` : "";
    return this.transport.request<SymbolsResponse>(
      `/api/v1/market/symbols${query}`,
      "GET",
      undefined,
      options
    );
  }

  /**
   * Retrieves current multi-asset price snapshot for all tracked global symbols.
   */
  public async getPrices(options?: RequestOptions): Promise<MarketPricesResponse> {
    const raw = await this.transport.request<MarketPricesResponse>(
        "/api/v1/market/prices",
        "GET",
        undefined,
        options
      );
      return { ...raw, items: Array.isArray(raw?.items) ? raw.items : [] };
  }

  /** Retrieves the latest quote for a single symbol. */
  public async getPrice(symbol: string, options?: RequestOptions): Promise<MarketPrice> {
    const cleanSymbol = normalizeSymbol(symbol);
    return this.transport.request<MarketPrice>(
      `/api/v1/market/prices/${encodeURIComponent(cleanSymbol)}`,
      "GET",
      undefined,
      options
    );
  }

  /** Retrieves the current exchange session state for a symbol. */
  public async getSession(symbol: string, options?: RequestOptions): Promise<unknown> {
    const cleanSymbol = normalizeSymbol(symbol);
    return this.transport.request<unknown>(
      `/api/v1/market/session/${encodeURIComponent(cleanSymbol)}`,
      "GET",
      undefined,
      options
    );
  }

  /** Retrieves data-quality and freshness diagnostics for market feeds. */
  public async getDataQuality(options?: RequestOptions): Promise<unknown> {
    return this.transport.request<unknown>("/api/v1/market/data-quality", "GET", undefined, options);
  }

  /** Retrieves detected short-window price spikes. */
  public async getSpikes(options?: RequestOptions): Promise<unknown> {
    return this.transport.request<unknown>("/api/v1/market/spikes", "GET", undefined, options);
  }

  /** Retrieves configured market alerts. */
  public async getAlerts(options?: RequestOptions): Promise<unknown> {
    return this.transport.request<unknown>("/api/v1/market/alerts", "GET", undefined, options);
  }

  /** Retrieves server-side smart alerts. */
  public async getSmartAlerts(options?: RequestOptions): Promise<unknown> {
    return this.transport.request<unknown>("/api/v1/market/smart-alerts", "GET", undefined, options);
  }

  /**
   * Fetches historical OHLCV candlestick bars for a given symbol and timeframe.
   *
   * @param symbol Alphanumeric instrument symbol (e.g. "XAUUSD", "BTCUSDT", "BBCA")
   * @param options Timeframe ("1m", "5m", "15m", "1h", "4h", "1d"), limit, timestamps
   */
  public async getCandles(
    symbol: string,
    options?: GetCandlesOptions
  ): Promise<CandleResponse> {
    const cleanSymbol = normalizeSymbol(symbol);
    const params = new URLSearchParams();
    if (options?.timeframe) params.set("resolution", options.timeframe);
    if (options?.limit) params.set("limit", String(options.limit));
    if (options?.since) params.set("since", String(options.since));
    if (options?.until) params.set("before", String(options.until));

    const query = params.toString() ? `?${params.toString()}` : "";
    const raw = await this.transport.request<{
      candles?: CandleResponse["candles"];
      items?: CandleResponse["candles"];
      has_more?: boolean;
      next_before?: number | null;
    }>(
      `/api/v1/market/history/${encodeURIComponent(cleanSymbol)}${query}`,
      "GET",
      undefined,
      options
    );

    const candleList = raw.candles || raw.items || [];
    return {
      symbol: cleanSymbol,
      timeframe: options?.timeframe || "1m",
      count: candleList.length,
      candles: candleList,
      has_more: raw?.has_more,
      next_before: raw?.next_before,
    };
  }

  /**
   * Retrieves Level 2 DOM (Depth of Market) order book for a symbol.
   */
  public async getOrderBook(symbol: string, options?: RequestOptions): Promise<OrderBook> {
    const cleanSymbol = normalizeSymbol(symbol);
    const raw = await this.transport.request<any>(
      `/api/v1/market/orderbook/${encodeURIComponent(cleanSymbol)}`,
      "GET",
      undefined,
      options
    );
    const data = raw?.data && typeof raw.data === "object" ? raw.data : raw;
    return {
      symbol: data?.symbol || cleanSymbol,
      bids: Array.isArray(data?.bids) ? data.bids : [],
      asks: Array.isArray(data?.asks) ? data.asks : [],
      timestamp: Number(data?.timestamp || Date.now()),
    };
  }

  /**
   * Retrieves AI/Quant market insights and price movement narrative for a given instrument.
   */
  public async getInsights(symbol: string, options?: RequestOptions): Promise<unknown> {
    const cleanSymbol = normalizeSymbol(symbol);
    return this.transport.request<unknown>(
      `/api/v1/market/insights/${encodeURIComponent(cleanSymbol)}`,
      "GET",
      undefined,
      options
    );
  }

  /**
   * Retrieves active market trading halts and circuit breaker triggers.
   */
  public async getTradingHalts(options?: RequestOptions): Promise<unknown> {
    return this.transport.request<unknown>(
      "/api/v1/market/trading-halts",
      "GET",
      undefined,
      options
    );
  }

  /**
   * Retrieves upcoming and historical corporate actions (dividends, splits, earnings).
   */
  public async getCorporateActions(options?: RequestOptions): Promise<unknown> {
    return this.transport.request<unknown>(
      "/api/v1/market/corporate-actions",
      "GET",
      undefined,
      options
    );
  }

  /**
   * Calculates historical realized volatility (HV) across rolling windows (10d, 30d, 90d).
   */
  public async getRealizedVolatility(symbol?: string, options?: RequestOptions): Promise<unknown> {
    const params = new URLSearchParams();
    if (symbol !== undefined) params.set("symbol", normalizeSymbol(symbol));
    const query = params.toString() ? `?${params.toString()}` : "";
    return this.transport.request<unknown>(
      `/api/v1/market/realized-volatility${query}`,
      "GET",
      undefined,
      options
    );
  }

  /**
   * Retrieves implied volatility (IV) surface and ATM volatility index.
   */
  public async getImpliedVolatility(symbol?: string, options?: RequestOptions): Promise<unknown> {
    const params = new URLSearchParams();
    if (symbol !== undefined) params.set("symbol", normalizeSymbol(symbol));
    const query = params.toString() ? `?${params.toString()}` : "";
    return this.transport.request<unknown>(
      `/api/v1/market/implied-volatility${query}`,
      "GET",
      undefined,
      options
    );
  }

  /**
   * @deprecated Use `getInsights(symbol)` instead.
   */
  public async getWhy(symbol: string, options?: RequestOptions): Promise<unknown> {
    return this.getInsights(symbol, options);
  }
}
